src/Microsoft.ML.Core/Utilities/FixedSizeQueue.cs

```diff
@@ -13,19 +13,30 @@ namespace Microsoft.ML.Runtime.Internal.Utilities
     ///  an item will result in discarding the least recently added item.
     /// </summary>
     [BestFriend]
-    internal sealed class FixedSizeQueue<T>
+    internal sealed class FixedSizeQueue<T> : ICloneable
     {
         private readonly T[] _array;
         private int _startIndex;
         private int _count;
 
+        public int StartIndex => _startIndex;
```

Comment:
>public int StartIndex => _startIndex; [](start = 8, length = 37)

Even though it's an internal class this doesn't mean that we should design it badly by exposing its guts. This makes it quite difficult to reason about.

As far as I can tell you just need this for serialization. What had happened if you'd just written the results of iterating from `i` over range of `0` to `q.Count` and writing `q[i]` when serializing, and in the reverse case just read and inserted the result of that writing back into the queue? Still would have worked as far as I can tell, would it not?

Reply:
I also need this to clone the state, which means I will need to know the start index, count and the content.

---
In reply to: [236475666](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236475666) [](ancestors = 236475666)

Reply:
The state had to be cloned in any case before your change came along. Somehow that managed to work without doing this.

Let's just be clear, I don't consider breaking something as basic as encapsulation for this structure to be acceptable. Please find another way.

---
In reply to: [236478429](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236478429) [](ancestors = 236478429,236475666)

Reply:
Agreed.

---
In reply to: [236482984](https://github.com/dotnet/machinelearning/pull/1727#discussion_r236482984) [](ancestors = 236482984,236478429,236475666)